### PR TITLE
[feat]: add new sample multirpc

### PR DESCRIPTION
**Pull request by Yan0613**
[feat]: add new sample multirpc

### CODE REVIEW

**Comment by DMwangnima** [comment]
Please change the title so that it can better describes this PR.

File: multirpc/go-client/cmd/main.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	svc, err := greet.NewGreetService(cliTriple)
+	if err != nil {
+		panic(err)
+	}
+
+	resp, err := svc.Greet(context.Background(), &greet.GreetRequest{Name: "hello world"})
+	if err != nil {
+		logger.Error(err)
+	}
+	logger.Infof("Greet multirpc response: %s", resp)
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:57]: You can change it to "Greet multirpc triple response: %s" for clarity.

File: multirpc/go-client/cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+	//test dubbo
+	ins_dubbo, err := dubbo.NewInstance(
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:60]: Actually, ```dubbo.Instance``` is global conception. Init once is okay.

File: multirpc/go-client/cmd/main.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		dubbo.WithRegistry(
+			registry.WithZookeeper(),
+			registry.WithAddress("127.0.0.1:2181"),
+		),
+	)
+	if err != nil {
+		panic(err)
+	}
+	//test triple
+	cliTriple, err := ins.NewClient(
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:42]: In this scenario, a client would create multi connections. So we just need to configure protocol with ```WithXXX``` in Reference layer.

File: multirpc/go-client/cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	cli_dubbo, err := ins_dubbo.NewClient(
+		client.WithClientProtocolDubbo(),
+		client.WithClientSerialization(constant.Hessian2Serialization),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	connDubbo, err := cli_dubbo.Dial("GreetProvider")
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:78]: Configure ```WithProtocolDubbo``` here.

File: multirpc/go-client/cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	if err != nil {
+		panic(err)
+	}
+	var respDubbo string
+	if err := connDubbo.CallUnary(context.Background(), []interface{}{"hello", "new", "dubbo"}, &respDubbo, "Greet"); err != nil {
+		logger.Errorf("GreetProvider.Greet err: %s", err)
+		return
+	}
+	logger.Infof("Get Response: %s", respDubbo)
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:87]: ditto.

File: multirpc/go-client/cmd/main.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+	//test json rpc
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:89]: For jsonrpc part, ditto.

File: multirpc/go-server/cmd/main.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (srv *GreetMultiRPCServer) Greet(ctx context.Context, req *greet.GreetRequest) (*greet.GreetResponse, error) {
+	resp := &greet.GreetResponse{Greeting: req.Name}
+	return resp, nil
+}
+
+func main() {
+	ins, err := dubbo.NewInstance(
+		dubbo.WithName("dubbo_multirpc_server"),
```
Comment (DMwangnima) [inline multirpc/go-server/cmd/main.go:42]: For configuring multiple protocols in server side, please refer to ```dubbo.WithProtocol```. You should inject 3 ```dubbo.WithProtocol``` here.

File: multirpc/go-server/cmd/main.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+			registry.WithZookeeper(),
+			registry.WithAddress("127.0.0.1:2181"),
+		),
+	)
+	if err != nil {
+		panic(err)
+	}
+	srv, err := ins.NewServer(
+		server.WithServerProtocol(
+			protocol.WithDubbo(),
```
Comment (DMwangnima) [inline multirpc/go-server/cmd/main.go:53]: If you take a look at ```protocol.WithDubbo() protocol.WithTriple() ...```, you will find that only the last one would take effect.
You should refer to ```server.WithProtocolIDs```.

File: multirpc/go-client/cmd/main.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+	respTriple, err := svc.Greet(context.Background(), &greet.GreetRequest{Name: "hello world"})
+	if err != nil {
+		logger.Error(err)
+	}
+	logger.Infof("Greet triple response: %s", respTriple.Greeting)
+
+	//Duboo
+	cliDubbo, err := ins.NewClient(
+		client.WithClientProtocolTriple(),
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:62]: This should be ```client.WithClientProtocolDubbo()```.

File: multirpc/go-client/cmd/main.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	if err = connDubbo.CallUnary(context.Background(), []interface{}{"hello", "new", "dubbo"}, &respDubbo, "SayHello"); err != nil {
+		logger.Errorf("GreetProvider.Greet err: %s", err)
+		return
+	}
+	logger.Infof("Get dubbo Response: %s", respDubbo)
+
+	//JsonRpc
+	cliJsonRpc, err := ins.NewClient(
+		client.WithClientProtocolJsonRPC(),
+		client.WithClientSerialization(constant.Hessian2Serialization),
```
Comment (DMwangnima) [inline multirpc/go-client/cmd/main.go:83]: Actually, JsonRPC has determined that serialization protocol is Json.(You can refer to /protocol/jsonrpc/json.go 111 line for details).

File: multirpc/go-server/cmd/main.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+func main() {
+	//triple
+	ins, err := dubbo.NewInstance(
+		dubbo.WithName("dubbo_multirpc_triple_server"),
+		dubbo.WithRegistry(
+			registry.WithZookeeper(),
+			registry.WithAddress("127.0.0.1:2181"),
+		),
+		//dubbo.WithProtocol(protocol.WithDubbo()),
```
Comment (DMwangnima) [inline multirpc/go-server/cmd/main.go:53]: Actually, invoking ```dubbo.WithProtocol``` three times is correct. In each ```dubbo.WithProtocol```, you can specify a protocol and its detail(e.g. port listen on).

File: multirpc/go-server/cmd/main.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+		dubbo.WithProtocol(protocol.WithTriple()),
+		//dubbo.WithProtocol(protocol.WithJSONRPC()),
+	)
+	if err != nil {
+		panic(err)
+	}
+	//var protocols []string
+	//protocols = append(protocols, "tri", "dubbo", "jsonrpc")
+	srv, err := ins.NewServer(
+		server.WithServerProtocol(
```
Comment (DMwangnima) [inline multirpc/go-server/cmd/main.go:63]: As we have configured protocols in ```dubbo.Instance``` layer. We do not need to configure protocols here.

**Review by DMwangnima — COMMENTED** [review]
That's basically correct. Just fix these issues I remained and remove the redundant comments.

**Comment by DMwangnima** [comment]
BTW, you can describe what you did in each commit. Not just "[feat]: add new sample multirpc" since it does not help reviewers or yourself to figure out what this commit did.

**Comment by Yan0613** [comment]
OK, I have fixed the previously mentioned issues and added integration test.

**Comment by EndlessSeeker** [comment]
btw plz add a note about 'multirpc' in What It Contains of Readme file in the root directory.

**Comment by Yan0613** [comment]
> btw plz add a note about 'multirpc' in What It Contains of Readme file in the root directory.
ok, I had added it. plz review it again.

**Review by chickenlj — APPROVED** [review]
LGTM.

**Pull request merged**